src/EventListener/SessionListener.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCacheBundle package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCacheBundle\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\EventListener\SessionListener as BaseSessionListener;
+
+/**
+ * Decorates the default Symfony session listener.
+ *
+ * The default Symfony session listener automatically makes responses private
+ * in case the session was started. This kills the user context feature of
+ * FOSCache. We disable the default behaviour only if the user context header
+ * is part of the Vary headers to reduce the possible impacts on other parts
+ * of your application.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+final class SessionListener implements EventSubscriberInterface
+{
+    /**
+     * @var BaseSessionListener
+     */
+    private $inner;
+
+    /**
+     * @var string
+     */
+    private $userHashHeader;
+
+    /**
+     * @var array
+     */
+    private $userIdentifierHeaders;
+
+    public function __construct(BaseSessionListener $inner, string $userHashHeader, array $userIdentifierHeaders)
+    {
+        $this->inner = $inner;
+        $this->userHashHeader = strtolower($userHashHeader);
+        $this->userIdentifierHeaders = array_map('strtolower', $userIdentifierHeaders);
```

Comment:
given that this is executed on every single request, i think we should do this in the DI extension class instead.